File: 백준/Gold/19644. 좀비 떼가 기관총 진지에도 오다니/좀비 떼가 기관총 진지에도 오다니.py
import sys
input = sys.stdin.readline

N = int(input())
len, damage = map(int, input().split())
C = int(input())
arr = [0]
for _ in range(N):
    arr.append(int(input()))

psum = [0] * (N + 1)

for i in range(1, N+1):
    now = psum[i-1] - psum[max(0, i-len)]
    if arr[i] <= now + damage:
        psum[i] = psum[i-1] + damage
    else:
        if C:
            C -= 1
            psum[i] = psum[i-1]
        else:
            print("NO")
            sys.exit()

print("YES")

File: 백준/Gold/19644. 좀비 떼가 기관총 진지에도 오다니/좀비 떼가 기관총 진지에도 오다니.js
const input = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n")
const L = Number(input[0]) // 기관총 앞쪽 거리
const [ML, MK] = input[1].split(" ").map(Number) // 기관총의 유효 사거리와 각 1m당 살상력
let Cammo = Number(input[2]) // 지뢰 개수
const zombies = [0]
for (let i=3; i<3+L; i++) zombies.push(Number(input[i]))  // 각 좀비들의 체력
const damageArray = new Array(L+1).fill(0)

let flag = true // 생존 변수

for (let i=1; i<L+1; i++){
    const now = damageArray[i-1] - damageArray[Math.max(0, i-ML)]

    if (zombies[i] <= now + MK) damageArray[i] = damageArray[i-1] + MK
    else {
        if (Cammo){
            Cammo--
            damageArray[i] = damageArray[i-1]
        } else {
            flag = false
            break
        }
    }
}

if (flag) console.log('YES')
else console.log('NO')
